### ISSUE TITLE
win32com Excel functions fail after freezing

### ISSUE THREAD

**Issue by aliencaocao** [pyinstaller/pyinstaller#7898]
## Description of the issue

I have an app that works with Excel on Windows. When running from python.exe everything works, but after building with pyinstaller, it started to act weirdly. Specifically, when trying to copy a sheet from an excel file to another excel file, it fails with a meaningless error message. It seems like instead of pasting it to the designated excel file, it made a new excel file and pasted it there. I found this out when I made win32com display everything it's doing on screen.

### Context information (for bug reports)

* Output of `pyinstaller --version`: ```5.13.1```
* Version of Python: 3.9.13
* Platform: Windows 11 Pro Chinese
* How you installed Python: python.org
* Did you also try this on another platform? Does it work there?
This works with win32com so only can work on windows

* try the latest development version, using the following command:

```shell
pip install https://github.com/pyinstaller/pyinstaller/archive/develop.zip
```

* follow *all* the instructions in our "If Things Go Wrong" Guide
  (https://github.com/pyinstaller/pyinstaller/wiki/If-Things-Go-Wrong) and

### Make sure [everything is packaged correctly](https://github.com/pyinstaller/pyinstaller/wiki/How-to-Report-Bugs#make-sure-everything-is-packaged-correctly)

  * [x] start with clean installation
  * [x] use the latest development version
  * [x] Run your frozen program **from a command window (shell)** — instead of double-clicking on it
  * [x] Package your program in **--onedir mode**
  * [x] Package **without UPX**, say: use the option `--noupx` or set `upx=False` in your .spec-file
  * [x] Repackage you application in **verbose/debug mode**. For this, pass the option `--debug` to `pyi-makespec` or `pyinstaller` or use `EXE(..., debug=1, ...)` in your .spec file.

### A minimal example program which shows the error

```
import os
import sys
from datetime import datetime
from pathlib import Path
from tkinter import *
from tkinter import filedialog
from tkinter.ttk import *

import pandas as pd
from win32com.client import Dispatch


def get_res_path(relative_path):
    """ Get absolute path to resource, works for dev and for PyInstaller
     Relative path will always get extracted into root!"""
    base_path = getattr(sys, '_MEIPASS', os.path.dirname(__file__))
    if os.path.isfile(os.path.join(base_path, relative_path)):
        return os.path.join(base_path, relative_path)
    else:
        raise FileNotFoundError(f'Embedded file {os.path.join(base_path, relative_path)} is not found!')


def process():
    start_button.config(state=DISABLED)
    # init win32com Excel module
    xl = Dispatch("Excel.Application")
    xl.Visible = True

    df = pd.read_excel(InputFilePath.get(), header=0, dtype={'Date': datetime, 'Start Time': str, 'End Time': str})
    df.fillna('', inplace=True)

    # Init medical template
    medical_template_path = get_res_path('test.xlsx')
    medical_df_internal = pd.read_excel(medical_template_path, index_col=0, skiprows=1, header=0, sheet_name='Internal')
    medical_df_internal.columns = [c.replace('\n', ' ').strip() for c in medical_df_internal.columns]
    medical_df_internal.dropna(inplace=True)
    medical_df_external = pd.read_excel(medical_template_path, index_col=0, skiprows=1, header=0, sheet_name='External')
    medical_df_external.columns = [c.replace('\n', ' ').strip() for c in medical_df_external.columns]
    medical_df_external.dropna(inplace=True)

    # Saving medical
    with pd.ExcelWriter(MedicOutputFilePath.get()) as writer:
        medical_df_internal.to_excel(writer, sheet_name='Internal')
        medical_df_external.to_excel(writer, sheet_name='External')

    path1 = Path(medical_template_path).absolute()
    path2 = Path(MedicOutputFilePath.get()).absolute()

    wb1 = xl.Workbooks.Open(Filename=path1)
    wb2 = xl.Workbooks.Open(Filename=path2)

    # copy the 3rd and 4th sheet as those are instructions
    ws1 = wb1.Worksheets(3)
    ws1.Copy(After=wb2.Worksheets(2))
    ws2 = wb1.Worksheets(4)
    ws2.Copy(After=wb2.Worksheets(3))


root = Tk()

InputFrame = LabelFrame(root, padding=(10, 10), text='Input')
InputFrame.pack(expand=True, fill=X, side=TOP, pady=(10, 10))
InputFilePath = StringVar(value='entry.xlsx')
Label(InputFrame, text='Input file: ').pack(side=LEFT, padx=(10, 0))
InputFileTextbox = Entry(InputFrame, textvariable=InputFilePath, width=80)
InputFileTextbox.pack(expand=True, fill=X, side=LEFT, padx=(5, 5))
InputFileSelector = Button(InputFrame, text='Select File',
                           command=lambda: InputFilePath.set(filedialog.askopenfilename(title='Choose input file', filetypes=[('Excel Files', '.xlsx')])))
InputFileSelector.pack(side=LEFT, padx=(0, 10))

OutputFrame = LabelFrame(root, padding=(10, 10), text='Output')
OutputFrame.pack(expand=True, fill=X, side=TOP, pady=(10, 10))

MedicFrame = Frame(OutputFrame)
MedicFrame.pack(expand=True, fill=X, side=TOP, pady=(10, 10))
MedicOutputFilePath = StringVar(value='test.xlsx')
Label(MedicFrame, text='Medical Excel file: ').pack(side=LEFT, padx=(10, 0))
MedicOutputFileTextbox = Entry(MedicFrame, textvariable=MedicOutputFilePath, width=80)
MedicOutputFileTextbox.pack(expand=True, fill=X, side=LEFT, padx=(0, 10))

start_button = Button(root, text='Start', command=process)
start_button.pack(expand=True, fill=X, side=TOP, pady=(10, 10), padx=(10, 10))
progress_bar_var = IntVar(value=0)
progress_bar_var.trace('w', lambda *args: root.update())  # ensure it dont freeze
progress_bar = Progressbar(root, orient=HORIZONTAL, length=100, mode='determinate', variable=progress_bar_var)
progress_bar.pack(expand=True, fill=X, side=TOP, pady=(10, 10), padx=(10, 10))

mainloop()
```

### Stacktrace / full error message

```
Exception in Tkinter callback
Traceback (most recent call last):
  File "tkinter\__init__.py", line 1892, in __call__
  File "main.py", line 229, in process
    # Copy the first and 2nd row in Internal and External sheet with formats
  File "win32com\client\dynamic.py", line 224, in __call__
pywintypes.com_error: (-2147352567, '发生意外。', (0, None, None, None, 0, -2147352565), None)
```

Build command for test app:
`pyinstaller --log-level DEBUG -F --add-data "test.xlsx;." --exclude-module matplotlib --exclude-module sqlite3 --exclude-module jinja2 --exclude-module PIL --exclude-module scipy --exclude-module urllib3 --exclude-module llvmlite --exclude-module numba --distpath ./ --clean --noconfirm "main.py"`

**Comment by rokm** [comment]
Just to clarify - is this a regression in 5.13.1, or was it problem with 5.13.0 as well?

Also, can you provide an example input XLSX file, so I can try reproducing the problem on my side? Thanks!

**Comment by aliencaocao** [comment]
i tried in 5.13.1 pip wheel and latest develop which reports itself as 5.13.0, both same issue.
you can actually make the xlsx file yourself. just have 1 xlsx named test.xlsx and have 4 sheets inside (first 2 named Internal and External), can be empty sheets
then make another xlsx named entry.xlsx and have 2 sheets inside, named Internal and External
the script is supposed to copy the 3rd and 4th sheet from test to entry, but instead it just makes a new document (unnamed too) and copy the 3rd sheet there. Then the code reaches the line for copying the 4th sheet, it errors because it expected 3 existing sheets.

**Comment by aliencaocao** [comment]
hi any idea on this issue? tried on another windows 10 system and still same. If there's anything else I can help to test/provide, let me know

**Comment by rokm** [comment]
Nope. I seem to be getting the same behavior when I run your code as a python script. 

I created `test.xlsx` with worksheets: `Internal`, `External`, `MySheet1`, and `MySheet2`.

The `entry.xlsx` is actually not used anywhere - you read it and fill empty values in 
```
    df = pd.read_excel(InputFilePath.get(), header=0, dtype={'Date': datetime, 'Start Time': str, 'End Time': str})
    df.fillna('', inplace=True)
```
but that `df` is not used anywhere. So I commented out those two lines.

I also modified `MedicOutputFilePath = StringVar(value='test.xlsx')` to have the default output set to `test-output.xlsx`, to avoid overwriting the template `test.xlsx`.

Running the processing then opens both `tests.xlsx` and `test-output.xlsx`. The latter has `Internal` and `External` worksheets, which you copied over using `pandas`.

Then the first attempt at copying worksheet via win32com opens a new unnamed document and copies over `MySheet1`. The second copy attempt raises the same `pywintypes.com_error: (-2147352567, 'Exception occurred.', (0, None, None, None, 0, -2147352565), None)` that you reported.

**Comment by rokm** [comment]
FWIW, [MSDN docs for Worksheet.Copy](https://learn.microsoft.com/en-us/office/vba/api/excel.worksheet.copy) state: "Copies the sheet to another location in the current workbook or a new workbook." The way this is worded makes me wonder if copying between current and another open workbook is, in fact, supposed to work at all.

**Comment by aliencaocao** [comment]
It works for me in python though, and using a different method (which is to select all data and paste in a new sheet) works fine, so the cross-workbook functions should be working, even if its not a 'new workbook'.
I can email you the actual full code and the data files i use tmr, i suspect the sample file and code I gave here are not fully working as expected as I tried to quickly minimize it to a sample for bug reporting. Maybe something else in front of this code sample caused a change.
